File: android/app/src/main/java/com/metanav/app/engine/NavigationEngine.kt
package com.metanav.app.engine

import android.content.Context
import android.graphics.Bitmap
import android.util.Log
import com.metanav.app.frames.RgbFrame
import com.metanav.app.vision.DepthEstimator
import com.metanav.app.vision.ObjectDetector
import com.metanav.core.Advisory
import com.metanav.core.Detection
import com.metanav.core.FrameObservation
import com.metanav.core.ObstacleReasoner
import com.metanav.core.ReasonerConfig
import com.metanav.core.SceneState
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import java.io.Closeable
import java.util.concurrent.Executors
import java.util.concurrent.atomic.AtomicBoolean

/** Frame in, scene + advisory out. Runs the models on one background thread and drops frames while busy. */
class NavigationEngine(context: Context, config: ReasonerConfig) : Closeable {
    private val executor = Executors.newSingleThreadExecutor { r -> Thread(r, "metanav-vision").apply { priority = Thread.NORM_PRIORITY + 1 } }
    private val busy = AtomicBoolean(false)
    private val reasoner = ObstacleReasoner(config)
    private val depth: DepthEstimator? = runCatching { DepthEstimator(context) }.onFailure { Log.e(TAG, "Depth model unavailable", it) }.getOrNull()
    private val detector: ObjectDetector? = runCatching { ObjectDetector(context) }.onFailure { Log.w(TAG, "Object detector unavailable", it) }.getOrNull()

    private val _scene = MutableStateFlow(SceneState())
    val scene: StateFlow<SceneState> = _scene.asStateFlow()

    private val _advisories = MutableSharedFlow<Advisory>(extraBufferCapacity = 4)
    val advisories: SharedFlow<Advisory> = _advisories.asSharedFlow()

    private val _preview = MutableStateFlow<Bitmap?>(null)
    val preview: StateFlow<Bitmap?> = _preview.asStateFlow()

    private val _stats = MutableStateFlow(Stats())
    val stats: StateFlow<Stats> = _stats.asStateFlow()

    val hasDepthModel: Boolean get() = depth != null

    private var lastProcessedAt = 0L
    private var frameCounter = 0
    private var lastDetections: List<Detection> = emptyList()

    data class Stats(val processedFps: Float = 0f, val inferenceMs: Long = 0)

    fun updateConfig(config: ReasonerConfig) {
        executor.execute { reasoner.config = config }
    }

    fun reset() {
        executor.execute { reasoner.reset(); _scene.value = SceneState() }
    }

    /** Non-blocking: schedules the frame unless the previous one is still being processed. */
    fun submit(frame: RgbFrame) {
        _preview.value = frame.bitmap
        val now = System.currentTimeMillis()
        if (now - lastProcessedAt < MIN_INTERVAL_MS) return
        if (!busy.compareAndSet(false, true)) return
        lastProcessedAt = now
        executor.execute {
            try { process(frame) } catch (t: Throwable) { Log.e(TAG, "frame processing failed", t) } finally { busy.set(false) }
        }
    }

    private fun process(frame: RgbFrame) {
        val started = System.currentTimeMillis()
        val depthMap = depth?.estimate(frame.bitmap)
        frameCounter++
        // The detector is the slower, less important model: run it every other frame.
        val det = detector
        if (det != null && frameCounter % 2 == 0) {
            lastDetections = runCatching { det.detect(frame.bitmap) }.getOrDefault(lastDetections)
        }
        val out = reasoner.process(FrameObservation(frame.timestampMs, depthMap, lastDetections))
        _scene.value = out.scene
        out.advisory?.let { _advisories.tryEmit(it) }
        val elapsed = System.currentTimeMillis() - started
        val prev = _stats.value
        val fps = 1000f / (System.currentTimeMillis() - started).coerceAtLeast(MIN_INTERVAL_MS)
        _stats.value = Stats(processedFps = 0.8f * prev.processedFps + 0.2f * fps, inferenceMs = elapsed)
    }

    override fun close() {
        executor.shutdownNow()
        depth?.close()
        detector?.close()
    }

    companion object {
        private const val TAG = "Metanav:Engine"
        /** Cap processing at ~8 fps: enough for walking speed, kind to the battery. */
        private const val MIN_INTERVAL_MS = 120L
    }
}

File: ios/Metanav/Engine/NavigationEngine.swift
import CoreImage
import Foundation
import MetanavCore
import UIKit

/// Frame in, scene + advisory out. Runs the models on one serial queue and drops frames while busy.
final class NavigationEngine {
    struct Stats { var processedFps: Float = 0; var inferenceMs: Int = 0 }

    var onScene: ((SceneState) -> Void)?
    var onAdvisory: ((Advisory) -> Void)?
    var onPreview: ((UIImage) -> Void)?
    var onStats: ((Stats) -> Void)?

    private let queue = DispatchQueue(label: "metanav.vision", qos: .userInitiated)
    private let reasoner: ObstacleReasoner
    private let depth: DepthEstimator
    private let detector: ObjectDetector?
    private let ciContext = CIContext()
    private var busy = false
    private let busyLock = NSLock()
    private var lastProcessedAt: Int64 = 0
    private var frameCounter = 0
    private var lastDetections: [Detection] = []
    private var stats = Stats()

    /// Cap processing at ~8 fps: enough for walking speed, kind to the battery.
    private let minIntervalMs: Int64 = 120

    init(config: ReasonerConfig) throws {
        reasoner = ObstacleReasoner(config: config)
        depth = try DepthEstimator()
        detector = try? ObjectDetector()
    }

    func updateConfig(_ config: ReasonerConfig) {
        queue.async { self.reasoner.config = config }
    }

    func reset() {
        queue.async { self.reasoner.reset() }
    }

    /// Non-blocking: schedules the frame unless the previous one is still being processed.
    func submit(_ frame: CapturedFrame) {
        let now = nowMs()
        if now - lastProcessedAt < minIntervalMs { return }
        busyLock.lock()
        if busy { busyLock.unlock(); return }
        busy = true
        busyLock.unlock()
        lastProcessedAt = now
        queue.async { [self] in
            defer { busyLock.lock(); busy = false; busyLock.unlock() }
            process(frame)
        }
    }

    private func process(_ frame: CapturedFrame) {
        let started = nowMs()
        let depthMap: DepthMap?
        do { depthMap = try depth.estimate(frame.image) } catch {
            NSLog("[Metanav] depth failed: \(error.localizedDescription)")
            depthMap = nil
        }
        frameCounter += 1
        // The detector matters less than depth: run it every other frame.
        if let detector, frameCounter % 2 == 0 {
            lastDetections = (try? detector.detect(frame.image)) ?? lastDetections
        }
        let output = reasoner.process(FrameObservation(timestampMs: frame.timestampMs, depth: depthMap, detections: lastDetections))
        onScene?(output.scene)
        if let advisory = output.advisory { onAdvisory?(advisory) }
        if let onPreview, let image = previewImage(frame.image) { onPreview(image) }

        let elapsed = Int(nowMs() - started)
        let fps = 1000 / Float(max(Int64(elapsed), minIntervalMs))
        stats = Stats(processedFps: 0.8 * stats.processedFps + 0.2 * fps, inferenceMs: elapsed)
        onStats?(stats)
    }

    private func previewImage(_ image: FrameImage) -> UIImage? {
        switch image {
        case .cgImage(let cg): return UIImage(cgImage: cg)
        case .pixelBuffer(let pb):
            let ci = CIImage(cvPixelBuffer: pb)
            guard let cg = ciContext.createCGImage(ci, from: ci.extent) else { return nil }
            return UIImage(cgImage: cg)
        }
    }
}
